### ISSUE TITLE
Revise Contexts

### ISSUE THREAD

**Issue by mecha** [Dhii/evaluable-interface#3]
Currently, contexts are [type hinted](https://github.com/Dhii/evaluable-interface/blob/master/src/EvaluableInterface.php#L23) as `ValueAwareInterface`. This ought to be revised.

`ContainerInterface` [has been proposed](https://github.com/Dhii/expression-interface/pull/11#issuecomment-356292209) as an alternative, making all contexts "composite". I agree with this to a certain degree. Below are my thoughts on this.

----

We should probably start by thinking about contexts in terms of call-time configuration and stop thinking about expressions. Firstly, this package does not mention expressions. Secondly, it makes our decisions that much easier to make.

To me, a context is some data set provided to the evaluable object _during_ evaluation, unlike the evaluable's state configuration which is determined during its instantiation. The most probable common two use cases for contexts (IMO) is to (a) provide values that will be included in the calculation of the result or (b) to control the calculation mechanism.

Right off the bat, I feel like its just natural to normalize contexts to always be able to hold more than one value. Previously in `dhii/expression-interface`, we had two types of contexts: regular and composite. This can get confusing, since the most generic interface will be the one to be type-hinted, which can lead to many implementations having to manually validate for other interfaces. Thus becoming an implementation burden.

Then there's the word "config" in the previous paragraph, which from a dependent's point of view is almost always synonymous to "container", at least if you're Dhii enough. `ContainerInterface` is usually our go-to solution for things like these.

However, `ContainerInterface` implies that the dependent object _knows_ what configuration it will receive, which might not always be the case. Being able to iterate over the contents of the context is probably a need that will arise at some point. An argument can be made however, that if the dependent does not know the keys and wishes to iterate, it can instead know just _one_ key and expect it to map to a traversable.

----

Further discussion required.

**Comment by XedinUnknown** [comment]
Yes. This is something we had considered in the beginning, when planning for this.

My evaluable requires a list of values. Why must I know the key at which they lie? And how can I ensure, at least in documentation, that it will be the right type? It's not possible to type-hint a return value for a specific key.

On the other hand, what if the evaluable suddenly needs more than one piece of data? How does one piece get distinguished from the other?

---

The solution to always use a `ContainerInterface` seems flexible enough to me, and future-proof. If the problem of knowing the keys cannot be solved completely, we need to at least have a way to ensure some consistency. Here's one way.

Our algorithms never depend on concretions; only abstractions, such as interfaces. This means that if there is a specific evaluable that works in a specific way, there will be an interface for it. This interface can hold constants that identify the keys, so that they can be provided and tested consistently. Alternatively, a specialized context, a sub-type of `ContainerInterface`, may hold those keys. Perhaps both have uses, depending on the situation. Either way, the problem of consistency can be solved, albeit perhaps not in the most elegant way.
